File: src/Kelvin.Server/Features/Sensors/Sensors.cs
using Kelvin.Server.Models;

namespace Kelvin.Server.Features.Sensors
{
  public record SensorRequest(
    Guid? Id,
    string Name,
    string MacAddress,
    bool HasBattery,
    bool HasHumiditySensor,
    bool HasCO2Sensor,
    float TemperatureCOffset = 0,
    float HumidityPercentageOffset = 0,
    short CO2LevelPpmOffset = 0
  );

  public record SensorResponse(
    Guid Id,
    string? Name,
    string? MacAddress,
    bool HasBattery,
    bool HasHumiditySensor,
    bool HasCO2Sensor,
    bool Enabled,
    float TemperatureCOffset = 0,
    float HumidityPercentageOffset = 0,
    short CO2LevelPpmOffset = 0,
    DateTimeOffset? LastReadingAt = null
  )
  {
    public static SensorResponse FromSensor(Sensor sensor)
    {
      return new SensorResponse(
        sensor.Id,
        sensor.Name,
        sensor.MacAddress,
        sensor.HasBattery,
        sensor.HasHumiditySensor,
        sensor.HasCO2Sensor,
        sensor.Enabled,
        sensor.TemperatureCOffset,
        sensor.HumidityPercentageOffset,
        sensor.CO2LevelPpmOffset
      );
    }
  }
}

File: src/Kelvin.Client/app/models/sensors.ts
export type Sensors = {
  sensors: Sensor[];
};

export type Sensor = {
  id: string;
  name: string;
  macAddress: string;
  hasBattery: boolean;
  hasCO2Sensor: boolean;
  hasHumiditySensor: boolean;
  enabled: boolean;
};

export type EnvironmentReading = {
  timestamp: string;
  temperatureC: number;
  humidityPercentage: number;
  cO2LevelPpm: number;
  areas: Record<string, SensorReading>;
};

export type SensorReading = {
  sensorId: string;
  temperatureC: number;
  humidityPercentage: number;
  cO2LevelPpm: number;
  batteryLevelPercentage?: number;
  createdAt: string;
  updatedAt: string;
};
